rinja/src/filters/escape.rs

```diff
@@ -160,6 +161,276 @@ pub trait Escaper: Copy {
+pub struct AutoEscaper<'a, T: fmt::Display + ?Sized, E: Escaper> {
+    text: &'a T,
+    escaper: E,
+}
+
+impl<'a, T: fmt::Display + ?Sized, E: Escaper> AutoEscaper<'a, T, E> {
+    #[inline]
+    pub fn new(text: &'a T, escaper: E) -> Self {
+        Self { text, escaper }
+    }
+}
+
+/// Use the provided escaper
+impl<'a, T: fmt::Display + ?Sized, E: Escaper> AutoEscape for &&AutoEscaper<'a, T, E> {
+    type Escaped = EscapeDisplay<&'a T, E>;
+    type Error = Infallible;
+
+    #[inline]
+    fn rinja_auto_escape(&self) -> Result<Self::Escaped, Self::Error> {
+        Ok(EscapeDisplay(self.text, self.escaper))
+    }
+}
+
+/// Types that implement this marker trait don't need to be HTML escaped
+///
+/// Please note that this trait is only meant as speed-up helper. In some odd circumcises rinja
+/// might still decide to HTML escape the input, so if this must not happen, then you need to use
+/// the [`|safe`](super::safe) filter to prevent the auto escaping.
+///
+/// If you are unsure if your type generates HTML safe output in all cases, then DON'T mark it.
+/// Better safe than sorry!
+pub trait HtmlSafe: fmt::Display {}
+
+impl<T: HtmlSafe + ?Sized> HtmlSafe for &T {}
+
+/// Don't escape HTML safe types
+impl<'a, T: HtmlSafe + ?Sized> AutoEscape for &AutoEscaper<'a, T, Html> {
+    type Escaped = &'a T;
+    type Error = Infallible;
+
+    #[inline]
+    fn rinja_auto_escape(&self) -> Result<Self::Escaped, Self::Error> {
+        Ok(self.text)
+    }
+}
+
+/// Mark the output of a filter as "maybe safe"
+///
+/// This struct can be used as a transparent return type of custom filters that want to mark
+/// their output as "safe" depending on some circumstances, i.e. that their output maybe does not
+/// need to be escaped.
+///
+/// If the filter is not used as the last element in the filter chain, then any assumption is void.
+/// Let the next filter decide if the output is safe or not.
+///
+/// ## Example
+///
+/// ```rust
+/// mod filters {
+///     use rinja::{filters::MaybeSafe, Result};
+///
+///     // Do not actually use this filter! It's an intentionally bad example.
+///     pub fn backdoor(s: impl ToString, enable: &bool) -> Result<MaybeSafe<String>> {
+///         Ok(MaybeSafe {
+///             text: s.to_string(),
+///             needs_escaping: !*enable,
+///         })
+///     }
+/// }
+///
+/// #[derive(rinja::Template)]
+/// #[template(
+///     source = "<div class='{{ klass|backdoor(enable_backdoor) }}'></div>",
+///     ext = "html"
+/// )]
+/// struct DivWithBackdoor<'a> {
+///     klass: &'a str,
+///     enable_backdoor: bool,
+/// }
+///
+/// assert_eq!(
+///     DivWithBackdoor { klass: "<script>", enable_backdoor: false }.to_string(),
+///     "<div class='&#60;script&#62;'></div>",
+/// );
+/// assert_eq!(
+///     DivWithBackdoor { klass: "<script>", enable_backdoor: true }.to_string(),
+///     "<div class='<script>'></div>",
+/// );
+/// ```
+pub struct MaybeSafe<T: fmt::Display> {
```

Comment:
I think this type would be better as an enum (with variables `Safe` and `NeedsEscaping`?).

Reply:
Yes, you're right. :+1: It's less error prone. And actually its easier to use I just noticed while implementing the change.